### ISSUE TITLE
Exalted system logic roll - possible API enhancement

### ISSUE THREAD

**Issue by Kilitar** [GreenImp/rpg-dice-roller#198]
Hello,
My RPG use similar system like games based on Exalted RPG system rolls. 

The core of this system is standard "D10 dice pool" against target number 
Something what can be flawlessly done using your api with syntax  "5d10>5".

Only difference is - if you roll 10, you get 2 sucsess points instead of 1.

Rolls using 5d10>5:           3-7*-6*-10*-1  = 3 sucssesses
Result I am looking for:    3-7*-6*-10**-1 = 4 sucssesses

Any chance to expand your system to cover Exalted RPG engines in the future? 

Thank you.

**Comment by GreenImp** [comment]
Hi @Kilitar thanks for the suggestion!

I had a similar request a few years ago, #22. From that, success pools were implemented, but the double 10 wasn't, as it seemed a bit tricky to do. The library has advanced a lot since, so it should be easier to do now.

I've been trying to find some good information on the Exalted rules around the "10 counts as double", to see if it could also apply to other dice rolls, but it's proven difficult to find much. I have seen hints that it can sometimes apply to 7, 8, and 9 as well.
Do you have any references that you could link to / share with me?

What I'm trying to figure out is what exact functionality is required.
My current thoughts are that it should allow a [compare point](https://greenimp.github.io/rpg-dice-roller/guide/notation/modifiers.html#compare-point) to be specified, so it's not restricted to the number 10 (Or max value of the die).
I'm not sure if it should allow specifying the value to increase by, or whether it should always be double.

Some thoughts:

**A specific "double" modifier (`dbl` / `dbl{cp}`)**
```js
// greater than 5 is success (`>5`), max roll (10) is double (`dbl`)
5d10>5dbl: [3, 7*, 6*, 10*dbl, 1] = 4

// greater than 5 is success (`>5`), greater than or equal to 7 is double (`dbl>=7`)
5d10>5dbl>=7: [3, 7*dbl, 6*, 10*dbl, 1] = 5
```

**A generic "adjust" (Or similar name) modifier**
_Not sure of a good syntax for this, so this wouldn't be final._
_Maybe something like `adj{cp}:{operation}`, where the operation is the equation to change the value by._
```js
// greater than 5 is success (`>5`), multiply values of 10 by 2
5d10>5adj=10:*2: [3, 7*, 6*, 10*, 1] = 4

// greater than 5 is success (`>5`), multiply values greater than or equal to 7 by 2
5d10>5adj>=7:*2: [3, 7*, 6*, 10*, 1] = 5
```

The benefit of the second method is that it's a lot more flexible (Although I'm not sure what the output should look like).
The downside is that it's quite long and could be confusing.

If you're able to provide any more information on the exact rules, then I can try and figure out exactly how it will need to work.

**Comment by GreenImp** [comment]
Just thought of another option for possible notation.
It could be a multiply modifier, where you specify the amount to multiply by, and an optional [compare point](https://greenimp.github.io/rpg-dice-roller/guide/notation/modifiers.html#compare-point).

**Multiply modifier (`mul{n}` / `mul{n}{cp}`)**
```js
// greater than 5 is success (`>5`), max roll (10) is multiplied by 2
5d10>5mul2: [3, 7*, 6*, 10*mul2, 1] = 4

// greater than 5 is success (`>5`), greater than or equal to 7 is multiplied by 2
5d10>5mul2>=7: [3, 7*dbl, 6*, 10*mul2, 1] = 5
```

I'm not overly happy with the suggested output in any of the options; `10*mul2`, `10*dbl` both seem a bit verbose.

**Comment by Kilitar** [comment]
Yeah, you are right, I did not want to overwhelm the suggestion with all available options of Exalted rules and other similar systems and rules expansions. There are systems (like we used) which use only basic rules and do not implement anything "extended. Like in my case double for 10's is just what I need.

**Basic:** 
D10 dice pool agains target number(TN). Everything above TN count as 1, but 10 count as  2 succsesses.

**Extensions** (for some Exalted-clone systems)
Lucky numbers: (for example 7) might count as double too, like if it was 10. Or like you pointed above (double everything above 7)
To not be so easy, the sucssess multiplier might be more than *2, it might be *3, *4 etc. 
Another "tricky" point is - if lucky number is BELOW TN, it does not count at all. 
(So if you roll 3D10 against TN 8 and have lucky number 7(*3), it does not matter. As in this case - 7 is bellow TN, it is allwayz fail and you get 0 from it).

Unlucky numbers:
Works very similar (and usually is applied to roll 1). Effect is, it deduct -1 (or multiplier) from total sucssesses.

To cover such complex "mismatch" like: 10d10 vs TN5, with 1 as umlucky -1, 10 as double, and 7 as lucky*3 would be something like : 10d10>5{1=-1;7=3;10=2}

I think optimal solution to cover all possibilities is just above. Allow to specify "dataset" of how to multiply each result. But it is long and might be confusing :( and in extreme case it may end like this: :P 
10d10>5{1mul-3;2mul-2;3mul-1;4mul-1;7mul3;10mul2}
Which should result as rolls:
1 =        -3
2 =        -2
3,4 =     -1
5 =         0
6,8,9 = +1
7 =       +3
10 =     +2
-------------------------------------------------------------------------------------

Or just stick with very basics (which should cover most of cases, in system used by me all of them) - double for top roll (6 for d6, 10 for d10), and consider add option to deduct -1 for 1 roll (fail)
5d10>5d    10=double sucs 
5d10>5f       1 =fail (-1) sucs
5d10>5df   (both above apply 10 for double, 1 for fail-1)

**Comment by GreenImp** [comment]
Hey, thanks for the feedback.

The Unlucky numbers are already possible with [success](https://greenimp.github.io/rpg-dice-roller/guide/notation/modifiers.html#target-success-dice-pool-cp) / [failure](https://greenimp.github.io/rpg-dice-roller/guide/notation/modifiers.html#target-failures-dice-pool-f-cp) modifiers.

So using your example above:

> 10d10 vs TN5, with 1 as unlucky -1, 10 as double, and 7 as lucky*3 would be something like : 10d10>5{1=-1;7=3;10=2}

```js
// anything greater than 5 is a success (+1), equal to 1 is failure (-1), everything else is zero
// Rolls; 1 = -1, 10 = 1, 2 = 0, 7 = 1, 9 = 1
10d10>5f=1: [1_, 10*, 2, 7*, 9*, 5, 3, 8*, 1_, 4] = 2
```

Just need to figure out the best way of handling the multiplication of numbers.

A multiplication modifier could work. Something like:

```js
// max roll on die is multiplied by 2
10d10mul2

// equal to 10 is multiplied by 3
10d10mul3=10

// equal or greater than 7 is multiplied by 2
10d10mul2>=7
```

It could be tricky to get it to work with when using success modifiers, and without (So it works with normal rolls as well), but it should be possible.

I'm thinking of changing the notation output for [success](https://greenimp.github.io/rpg-dice-roller/guide/notation/modifiers.html#target-success-dice-pool-cp) / [failure](https://greenimp.github.io/rpg-dice-roller/guide/notation/modifiers.html#target-failures-dice-pool-f-cp) modifiers, so instead of `*` for success, and `_` for failure, it would be `+`, and `-`.
That way, the notation for the multiplication modifier could be something like `*2`. e.g.:

```js
5d10>=7f=1mul2: [4, 8+, 1-, 10+*2, 2] = 2
```

Does that seem sensible, and cover the basics?

**Comment by Kilitar** [comment]
Yep, that covers the basics and yes, use + for success,  - for failure and * for multiply looks like a good idea :).

**Comment by GreenImp** [comment]
Okay, thanks for the feedback. I'll look at implementing something like this when I can.

PRs are always welcome as well :)

**Comment by GreenImp** [comment]
Sorry for the delay. I've added a Multiply modifier, which is curently in the `develop` branch. I think it does what you need.

I'll be looking at getting it out in v5.0.0 soon.

**Comment by GreenImp** [comment]
I've had to remove this from the `develop` branch because it's having some conflicts with other notations, in terms of the visual flags it adds to the results.

The functionality is working though. I'm going to have a think about how best to handle the conflicts.

**Comment by nico198x** [comment]
> I've had to remove this from the `develop` branch because it's having some conflicts with other notations, in terms of the visual flags it adds to the results.
> 
> The functionality is working though. I'm going to have a think about how best to handle the conflicts.

Following up again on this. it's been 5 years and i could really use this functionality for Curseborne and the Storypath Ultra system (double 10s and double 9s).

Any new thoughts on this and how we can complete this? it seemed you were close before. what's holding you back? is there anything i can do to help facilitate?
